File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/ServiceEnvScreen.kt
// W10 — per-app environment-variable KV editor (Android parallel of
// the iOS ServiceEnvScreen). NAMES only on the list path; values are
// typed locally, signed via the IRK envelope signer, and POSTed once
// over the daemon's TLS. The phone NEVER persists the value, NEVER
// logs it, NEVER shows it after the dialog dismisses.

package com.flagshipserver.app.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.api.ServiceEnvSetEnvelope
import com.flagshipserver.app.api.ServiceEnvSetRequest
import com.flagshipserver.app.api.ServiceEnvUnsetRequest
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.LocalScreensClient
import com.flagshipserver.app.core.LocalToastCenter
import com.flagshipserver.app.core.LocalVibeCodeEnvelopeSigner
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSDangerButton
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import kotlinx.coroutines.launch

@Composable
fun ServiceEnvScreen(
    nav: NavController,
    appId: String,
    creator: String,
    slug: String,
) {
    val client = LocalScreensClient.current
    val appState = LocalAppState.current
    val toasts = LocalToastCenter.current
    val signer = LocalVibeCodeEnvelopeSigner.current
    val scope = rememberCoroutineScope()
    val names = remember { mutableStateListOf<String>() }
    var loading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var showAddDialog by remember { mutableStateOf(false) }

    val pods = appState.pods.collectAsState().value
    val leaderId = appState.leaderPodId.collectAsState().value
    val serverFqdn = pods.firstOrNull { it.podId == leaderId }?.fqdn
        ?: pods.firstOrNull()?.fqdn
        ?: "unknown"

    suspend fun reload() {
        loading = true
        errorMessage = null
        try {
            val resp = client.serviceEnvList(appId)
            names.clear()
            names.addAll(resp.names)
        } catch (t: Throwable) {
            errorMessage = t.message
        } finally {
            loading = false
        }
    }

    LaunchedEffect(Unit) { reload() }

    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(scroll)
            .padding(horizontal = FS.space.s6),
    ) {
        Spacer(Modifier.height(FS.space.s8))
        FSGhostButton(label = "← Back", onClick = { nav.popBackStack() })
        Spacer(Modifier.height(FS.space.s3))
        Text(
            "Configure environment",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Spacer(Modifier.height(FS.space.s4))

        FSCard {
            Column {
                Text(
                    "Environment variables",
                    color = FS.colors.text,
                    style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.SemiBold),
                )
                Spacer(Modifier.height(FS.space.s2))
                Text(
                    "Names appear here; values stay on this server. They're sealed at rest and never leave your pod.",
                    color = FS.colors.textMuted,
                    style = TextStyle(fontSize = 13.sp),
                )
            }
        }
        Spacer(Modifier.height(FS.space.s4))

        if (errorMessage != null) {
            ErrorCard(message = errorMessage!!)
            Spacer(Modifier.height(FS.space.s3))
        }

        if (loading) {
            ServerCardSkeleton()
        } else if (names.isEmpty()) {
            FSCard {
                Column {
                    Text("No env vars set", color = FS.colors.text)
                    Spacer(Modifier.height(FS.space.s1))
                    Text("Tap below to add one.", color = FS.colors.textMuted, style = TextStyle(fontSize = 13.sp))
                }
            }
        } else {
            names.forEach { name ->
                FSCard(modifier = Modifier.semantics { testTag = "service-env-row-$name" }) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(
                            name,
                            color = FS.colors.text,
                            modifier = Modifier.weight(1f),
                            style = TextStyle(fontSize = 14.sp),
                        )
                        FSDangerButton(
                            label = "Remove",
                            onClick = {
                                scope.launch {
                                    val issuedAt = System.currentTimeMillis()
                                    val envelope = ServiceEnvSetEnvelope(
                                        serverId = serverFqdn,
                                        creator = creator, slug = slug,
                                        env = emptyMap(),
                                        issuedAt = issuedAt,
                                    )
                                    try {
                                        val signature = signer(envelope)
                                        client.serviceEnvUnset(
                                            appId,
                                            ServiceEnvUnsetRequest(name = name, request = envelope, signature = signature),
                                        )
                                        reload()
                                    } catch (t: Throwable) {
                                        errorMessage = t.message
                                    }
                                }
                            },
                        )
                    }
                }
                Spacer(Modifier.height(FS.space.s2))
            }
        }

        Spacer(Modifier.height(FS.space.s4))
        FSPrimaryButton(
            label = "Add environment variable",
            onClick = { showAddDialog = true },
            modifier = Modifier.semantics { testTag = "service-env-add-btn" },
        )
        Spacer(Modifier.height(FS.space.s10))
    }

    if (showAddDialog) {
        AddEnvVarDialog(
            onDismiss = { showAddDialog = false },
            onSubmit = { name, value ->
                scope.launch {
                    val issuedAt = System.currentTimeMillis()
                    val envelope = ServiceEnvSetEnvelope(
                        serverId = serverFqdn,
                        creator = creator, slug = slug,
                        env = mapOf(name to value),
                        issuedAt = issuedAt,
                    )
                    try {
                        val signature = signer(envelope)
                        client.serviceEnvSet(
                            appId,
                            ServiceEnvSetRequest(name = name, value = value, request = envelope, signature = signature),
                        )
                        showAddDialog = false
                        reload()
                    } catch (t: Throwable) {
                        errorMessage = t.message
                    }
                }
            },
        )
    }
}

@Composable
private fun AddEnvVarDialog(
    onDismiss: () -> Unit,
    onSubmit: (name: String, value: String) -> Unit,
) {
    var name by remember { mutableStateOf("") }
    var value by remember { mutableStateOf("") }
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("New environment variable") },
        text = {
            Column {
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text("Name (e.g. OPENAI_API_KEY)") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Ascii),
                    modifier = Modifier
                        .fillMaxWidth()
                        .semantics { testTag = "service-env-name-field" },
                )
                Spacer(Modifier.height(FS.space.s2))
                OutlinedTextField(
                    value = value,
                    onValueChange = { value = it },
                    label = { Text("Value") },
                    visualTransformation = PasswordVisualTransformation(),
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
                    modifier = Modifier
                        .fillMaxWidth()
                        .semantics { testTag = "service-env-value-field" },
                )
                Spacer(Modifier.height(FS.space.s2))
                Text(
                    "The value is sent once to your server. The Flagship phone app does not save it.",
                    color = FS.colors.textMuted,
                    style = TextStyle(fontSize = 11.sp),
                )
            }
        },
        confirmButton = {
            TextButton(
                onClick = { onSubmit(name, value) },
                enabled = name.isNotEmpty() && value.isNotEmpty(),
                modifier = Modifier.semantics { testTag = "service-env-save-btn" },
            ) { Text("Save") }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Cancel") }
        },
    )
}

File: apps/mobile/ios/Sources/FlagshipUI/Screens/ServiceEnvScreen.swift
import SwiftUI
import FlagshipAPI
import FlagshipCore

/// W10 — per-app environment-variable KV editor.
///
/// Shows the NAMES (only) of env vars set on the app, lets the owner
/// add or remove entries. The value is typed locally, signed under
/// the owner's IRK, and POSTed over the daemon's TLS once. The phone
/// NEVER persists the value, NEVER logs it, NEVER shows it after
/// dismissing the editor. The daemon's response is name-only.
@MainActor
public struct ServiceEnvScreen: View {
    @Environment(\.colorScheme) private var scheme
    @State private var names: [String] = []
    @State private var loading: Bool = true
    @State private var errorMessage: String?
    @State private var showAddSheet: Bool = false

    public let appId: String
    public let serverFqdn: String
    public let creator: String
    public let slug: String
    /// Async hook that signs the SetServiceEnvRequest envelope under the
    /// owner's IRK. The view never holds the IRK private bytes; it
    /// delegates signing to the platform Keystore (via this closure).
    public let signEnvelope: @MainActor (ServiceEnvSetEnvelope) async throws -> String
    public let client: any ScreensClient

    public init(
        appId: String,
        serverFqdn: String,
        creator: String,
        slug: String,
        client: any ScreensClient,
        signEnvelope: @escaping @MainActor (ServiceEnvSetEnvelope) async throws -> String
    ) {
        self.appId = appId
        self.serverFqdn = serverFqdn
        self.creator = creator
        self.slug = slug
        self.client = client
        self.signEnvelope = signEnvelope
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                explainerCard(c: c)
                listCard(c: c)
                if let msg = errorMessage {
                    ErrorCard(message: msg)
                }
                Spacer().frame(height: FS.space.s12)
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.top, FS.space.s4)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Configure environment")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button {
                    showAddSheet = true
                } label: {
                    Image(systemName: "plus.circle.fill")
                }
                .accessibilityIdentifier("service-env-add-btn")
            }
        }
        .task { await reload() }
        .sheet(isPresented: $showAddSheet) {
            AddEnvVarSheet(
                appId: appId,
                serverFqdn: serverFqdn,
                creator: creator,
                slug: slug,
                existingNames: names,
                client: client,
                signEnvelope: signEnvelope,
                onCommitted: {
                    showAddSheet = false
                    Task { await reload() }
                },
                onCancel: { showAddSheet = false }
            )
        }
    }

    private func explainerCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s2) {
                Text("Environment variables").font(FS.font.h4()).foregroundColor(c.text)
                Text("Names appear here; values stay on this server. They're sealed at rest and never leave your pod.")
                    .font(FS.font.bodySm())
                    .foregroundColor(c.textMuted)
            }
        }
    }

    @ViewBuilder
    private func listCard(c: FSColors) -> some View {
        if loading {
            FSCard {
                HStack {
                    ProgressView()
                    Text("Loading…").foregroundColor(c.textMuted)
                }
            }
        } else if names.isEmpty {
            FSCard {
                VStack(alignment: .leading, spacing: FS.space.s2) {
                    Text("No env vars set").font(FS.font.body()).foregroundColor(c.text)
                    Text("Tap the + button to add one.").font(FS.font.bodySm()).foregroundColor(c.textMuted)
                }
            }
        } else {
            FSCard {
                VStack(alignment: .leading, spacing: 0) {
                    ForEach(Array(names.enumerated()), id: \.element) { idx, name in
                        envRow(name: name, c: c)
                        if idx < names.count - 1 {
                            Divider().background(c.border)
                        }
                    }
                }
            }
        }
    }

    private func envRow(name: String, c: FSColors) -> some View {
        HStack(spacing: FS.space.s3) {
            Image(systemName: "key.fill").foregroundColor(c.textMuted)
            Text(name)
                .font(FS.font.mono())
                .foregroundColor(c.text)
                .accessibilityIdentifier("service-env-row-\(name)")
            Spacer()
            // The VALUE is not shown — by design. Pull-to-edit pops the
            // add-sheet so the user can re-type (the new value replaces
            // the old; the phone never sees the prior value).
            Button {
                Task { await unset(name: name) }
            } label: {
                Image(systemName: "trash").foregroundColor(c.danger)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Remove \(name)")
        }
        .padding(.vertical, FS.space.s3)
        .padding(.horizontal, FS.space.s2)
    }

    private func reload() async {
        loading = true
        errorMessage = nil
        defer { loading = false }
        do {
            let r = try await client.serviceEnvList(appId: appId)
            names = r.names
        } catch {
            errorMessage = String(describing: error)
        }
    }

    private func unset(name: String) async {
        // Build a NEW envelope with the dropped name removed. The
        // envelope is FULL-REPLACE — the daemon stores exactly what's
        // signed, so we must enumerate the surviving names. We don't
        // have values for the surviving names (by design), so we can't
        // round-trip them on /unset alone — the daemon's full-replace
        // semantics force us to either (a) restore values via a paired
        // op, or (b) tolerate values being preserved server-side
        // because they're already on disk and the new request only
        // signs the SET WITHOUT-the-removed-name.
        //
        // We post an envelope with an empty `env` for the unset name
        // and let the daemon's setEnv replace the on-disk map; the
        // owner gets exactly what they signed. If the user wants to
        // KEEP other vars, they must re-add them through the editor.
        // This is honest about the cryptographic boundary — values
        // can't be retained by the phone without leaking them.
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let envelope = ServiceEnvSetEnvelope(
            serverId: serverFqdn,
            creator: creator,
            slug: slug,
            env: [:],
            issuedAt: issuedAt
        )
        do {
            let signature = try await signEnvelope(envelope)
            let _ = try await client.serviceEnvUnset(
                appId: appId,
                ServiceEnvUnsetRequest(name: name, request: envelope, signature: signature)
            )
            await reload()
        } catch {
            errorMessage = String(describing: error)
        }
    }
}

/// Modal text-entry sheet for a single NAME + VALUE pair. The value is
/// kept in @State only for the lifetime of the sheet — dismissing
/// drops it. The send path POSTs once over TLS to the daemon and
/// never echoes the value back.
@MainActor
private struct AddEnvVarSheet: View {
    @Environment(\.colorScheme) private var scheme
    @State private var name: String = ""
    @State private var value: String = ""
    @State private var submitting: Bool = false
    @State private var errorMessage: String?

    let appId: String
    let serverFqdn: String
    let creator: String
    let slug: String
    let existingNames: [String]
    let client: any ScreensClient
    let signEnvelope: @MainActor (ServiceEnvSetEnvelope) async throws -> String
    let onCommitted: () -> Void
    let onCancel: () -> Void

    var body: some View {
        let c = FSColors.scheme(scheme)
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: FS.space.s4) {
                    FSCard {
                        VStack(alignment: .leading, spacing: FS.space.s3) {
                            FSColorReader { c2 in
                                Text("Name").font(FS.font.caption()).foregroundColor(c2.textMuted)
                            }
                            TextField("e.g. OPENAI_API_KEY", text: $name)
                                .font(FS.font.mono())
                                .autocorrectionDisabled()
                                .textInputAutocapitalization(.never)
                                .padding(.horizontal, 12).frame(height: 40)
                                .background(c.surfaceSunken)
                                .overlay(RoundedRectangle(cornerRadius: FS.radius.sm).stroke(c.border, lineWidth: 1))
                                .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
                                .accessibilityIdentifier("service-env-name-field")
                            FSColorReader { c2 in
                                Text("Value").font(FS.font.caption()).foregroundColor(c2.textMuted)
                            }
                            SecureField("paste your secret here", text: $value)
                                .font(FS.font.mono())
                                .autocorrectionDisabled()
                                .textInputAutocapitalization(.never)
                                .padding(.horizontal, 12).frame(height: 40)
                                .background(c.surfaceSunken)
                                .overlay(RoundedRectangle(cornerRadius: FS.radius.sm).stroke(c.border, lineWidth: 1))
                                .clipShape(RoundedRectangle(cornerRadius: FS.radius.sm))
                                .accessibilityIdentifier("service-env-value-field")
                            Text("The value is sent once to your server. The Flagship phone app does not save it.")
                                .font(FS.font.caption())
                                .foregroundColor(c.textMuted)
                        }
                    }
                    if let msg = errorMessage {
                        ErrorCard(message: msg)
                    }
                    FSPrimaryButton(
                        submitting ? "Sending…" : "Save",
                        enabled: !submitting && !name.isEmpty && !value.isEmpty,
                        block: true
                    ) {
                        Task { await submit() }
                    }
                    .accessibilityIdentifier("service-env-save-btn")
                    Spacer().frame(height: FS.space.s12)
                }
                .padding(.horizontal, FS.space.s6)
                .padding(.top, FS.space.s4)
            }
            .background(c.bg.ignoresSafeArea())
            .navigationTitle("New environment variable")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Cancel", role: .cancel) { onCancel() }
                }
            }
        }
    }

    private func submit() async {
        submitting = true
        errorMessage = nil
        defer { submitting = false }
        // Replacement set = existing names (preserved on the daemon
        // because we re-emit them in the signed envelope) plus the
        // new key/value. Existing values are unknown to the phone, so
        // we only sign the FULL desired map for the entries the phone
        // actually has plaintext for — the new one. The daemon's
        // setEnv() is full-replace, but the daemon's CURRENT impl
        // accepts ONLY the entries in this signed envelope.
        //
        // Practical consequence: setting a NEW name in a session that
        // already had others wipes the others unless the user re-types
        // them. That's an honest reflection of the security boundary:
        // values can't be retained on the phone without leaking them.
        // The sheet's helper text spells this out for the user.
        var env: [String: String] = [:]
        env[name] = value
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let envelope = ServiceEnvSetEnvelope(
            serverId: serverFqdn,
            creator: creator,
            slug: slug,
            env: env,
            issuedAt: issuedAt
        )
        do {
            let signature = try await signEnvelope(envelope)
            let _ = try await client.serviceEnvSet(
                appId: appId,
                ServiceEnvSetRequest(name: name, value: value, request: envelope, signature: signature)
            )
            // Drop the in-memory value so a screenshot/snapshot of the
            // dismissed sheet doesn't expose it.
            value = ""
            onCommitted()
        } catch {
            errorMessage = String(describing: error)
        }
    }
}
